### DISCUSSION TITLE
Access dicom nested tags efficiently

### DISCUSSION THREAD

**Discussion by scandav** [Enet4/dicom-rs#334] [Q&A]
I am trying to access some nested tags in a dicom in an efficient way. Such tags could also be not present.

So far I have a basic implementation that does the job, but as you see it has a lot of `.unwrap()` calls. How to make it more efficient and handle the `Option` results all at once?

``` rust
    let obj = open_file("../dicoms/my_dicom_path.dcm")?;
    
    let referenced_sop_instance_uid = obj
        .element(tags::SHARED_FUNCTIONAL_GROUPS_SEQUENCE)?
        .items()
        .unwrap()
        .first()
        .unwrap()
        .element(tags::REFERENCED_IMAGE_SEQUENCE)?
        .items()
        .unwrap()
        .first()
        .unwrap()
        .element(tags::REFERENCED_SOP_INSTANCE_UID)?
        .to_str()?;
```

**Comment by Enet4** [comment]
Sorry, this question must have slipped. Yes, the errors need to be handled at each access level. Note that your code also panics if the DICOM object is malformed and the value behind a sequence tag is something other than a sequence. By using Snafu, both results and options can be handled appropriately. Here is a function that does better safeguarding (consider later on migrating from `Whatever` to your own error types).

```rust
use snafu::{ResultExt, OptionExt};

fn get_item<'a, D>(
    obj: &'a InMemDicomObject<D>,
    tag: Tag,
    name: &'static str,
) -> Result<&'a InMemDicomObject<D>, snafu::Whatever>
where
    D: Clone + DataDictionary,
{
    Ok(obj
        .element(tag)
        .whatever_context("missing attribute")?
        .items()
        .whatever_context("not a sequence")?
        .get(0)
        .context("item index in sequence out of range")?)
}
```

However, with the attribute selector API around the corner (#326 #383), we could also have a `value_at` method which traverses the DICOM tree for you in a single call, making these forms of access much easier.
